File: src/main/java/flat/viewer/web/FlatViewController.java
package flat.viewer.web;

import flat.viewer.FlatViewService;
import flat.viewer.NotificationService;
import flat.viewer.NotificationServiceImpl;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.Vertx;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.handler.BodyHandler;

import java.time.LocalTime;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

import static io.netty.handler.codec.http.HttpHeaders.Values.APPLICATION_JSON;
import static java.time.temporal.ChronoUnit.MINUTES;

//import io.vertx.rxjava.core.AbstractVerticle;
//import io.vertx.rxjava.core.http.HttpServerRequest;
//import io.vertx.rxjava.ext.web.Router;
//import io.vertx.rxjava.ext.web.RoutingContext;

public class FlatViewController extends AbstractVerticle {

    private static final AtomicLong seq = new AtomicLong();

    private static final String HOST = "localhost";
    private static final int PORT = 8080;

    public static void main(String[] args) {
        Vertx.vertx().deployVerticle(new FlatViewController());
    }

    @Override
    public void start() {
//        AuctionRepository repository = new AuctionRepository(vertx.sharedData());
//        FVValidator validator = new FVValidator(repository);
        NotificationService notificationService = new NotificationServiceImpl();
        FlatViewService flatViewService = new FlatViewService(notificationService);
        //        FVValidator validator = new FVValidator();
//        AuctionHandler handler = new AuctionHandler(repository, validator);
        LocalTime start = LocalTime.of(10, 0);
        LocalTime end = LocalTime.of(20, 0);
        int duration = 20;
        FlatViewHandler handler = new FlatViewHandler(flatViewService, initTimeSlots(start, end, duration));

        Router router = Router.router(vertx);
        router.route().handler(BodyHandler.create());
        router.route().consumes(APPLICATION_JSON);
        router.route().produces(APPLICATION_JSON);

        router.post("/rent").handler(handler::handleRent);
        router.post("/reserve").handler(handler::handleReserve);
        router.post("/cancel").handler(handler::handleCancel);
        router.post("/approve").handler(handler::handleApprove);
        router.post("/reject").handler(handler::handleReject);

        vertx.createHttpServer().requestHandler(router).listen(PORT, HOST);


//        SharedData sharedData = vertx.sharedData();
//
//        sharedData.getLockWithTimeout("mylock", 10000, res -> {
//            if (res.succeeded()) {
//                // Got the lock!
//                Lock lock = res.result();
//
//                LocalMap<String, String> map1 = sharedData.getClusterWideMap("mymap1");
//                
//            } else {
//                // Failed to get lock
//            }
//        });

    }

    private static Set<Map.Entry<LocalTime, LocalTime>> initTimeSlots(LocalTime start, LocalTime end, int duration) {
        long minutes = MINUTES.between(start, end);
        long count = minutes / duration;
        List<Map.Entry<LocalTime, LocalTime>> timeSlots = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            int add = duration * i;
            timeSlots.add(new AbstractMap.SimpleEntry<>(start.plus(add, MINUTES), start.plus(add + duration, MINUTES)));
        }
        long left = minutes % duration;
        if (left > 0) {
            Map.Entry<LocalTime, LocalTime> entry = timeSlots.get(timeSlots.size() - 1);
            timeSlots.add(new AbstractMap.SimpleEntry<>(entry.getValue(), entry.getValue().plus(left, MINUTES)));
        }
        return new HashSet<>(timeSlots);
    }
}

File: src/main/java/flat/viewer/web/FlatViewController.kt
package flat.viewer.web

import flat.viewer.FlatViewService
import flat.viewer.NotificationService
import flat.viewer.NotificationServiceImpl
import flat.viewer.TimeSlotsHelper
import io.netty.handler.codec.http.HttpHeaderValues
import io.vertx.core.AbstractVerticle
import io.vertx.core.Vertx
import io.vertx.ext.web.Router
import io.vertx.ext.web.handler.BodyHandler
import java.time.LocalTime

class FlatViewController : AbstractVerticle() {
    override fun start() {
        val notificationService: NotificationService = NotificationServiceImpl()
        val flatViewService = FlatViewService(notificationService)
        val start = LocalTime.of(10, 0)
        val end = LocalTime.of(20, 0)
        val duration = 20
        val handler = FlatViewHandler(flatViewService, TimeSlotsHelper.initTimeSlots(start, end, duration), vertx.sharedData())
        val router = Router.router(vertx)
        router.route().handler(BodyHandler.create())
        router.route().consumes(HttpHeaderValues.APPLICATION_JSON.toString())
        router.route().produces(HttpHeaderValues.APPLICATION_JSON.toString())
        router.post("/rent").handler(handler::handleRent)
        router.post("/reserve").handler(handler::handleReserve)
        router.post("/cancel").handler(handler::handleCancel)
        router.post("/approve").handler(handler::handleApprove)
        router.post("/reject").handler(handler::handleReject)
        vertx.createHttpServer().requestHandler(router).listen(PORT, HOST)
    }

    companion object {
        private const val HOST = "localhost"
        private const val PORT = 8080
        @JvmStatic
        fun main(args: Array<String>) {
            Vertx.vertx().deployVerticle(FlatViewController())
        }
    }
}
